lib/ldclient-rb/ldclient.rb

```diff
@@ -226,10 +244,116 @@ def variation(key, context, default)
     # @return [EvaluationDetail] an object describing the result
     #
     def variation_detail(key, context, default)
-      detail, _, _ = evaluate_internal(key, context, default, true)
+      detail, _, _ = evaluate_with_hooks(key, context, default, :variation_detail) do
+        evaluate_internal(key, context, default, true)
+      end
+
       detail
     end
 
+    #
+    # evaluate_with_hook will run the provided block, wrapping it with evaluation hook support.
+    #
+    # Example:
+    #
+    # ```ruby
+    # evaluate_with_hooks(key, context, default, method) do
+    #   puts 'This is being wrapped with evaluation hooks'
+    # end
+    # ```
+    #
+    # @param key [String]
+    # @param context [String]
+    # @param default [String]
+    # @param method [Symbol]
+    # @param &block [#call] Implicit passed block
+    #
+    private def evaluate_with_hooks(key, context, default, method)
+      return yield if @hooks.empty?
+
+      hooks, hook_context = prepare_hooks(key, context, default, method)
+      hook_data = execute_before_evaluation(hooks, hook_context)
+      evaluation_detail, flag, error = yield
+      execute_after_evaluation(hooks, hook_context, hook_data, evaluation_detail)
+
+      [evaluation_detail, flag, error]
+    end
+
+    #
+    # Execute the :before_evaluation stage of the evaluation series.
+    #
+    # This method will return the results of each hook, indexed into an array in the same order as the hooks. If a hook
+    # raised an uncaught exception, the value will be nil.
```

Comment:
Would this nil propagate to the after?

Reply:
Yes. The results of this method get returned as an array. We then use that array in the after_evaluation .zip call to paiar it with the appropriate hook.